core/services/eth/client.go

```diff
@@ -256,7 +250,81 @@ func (client *client) SubscribeFilterLogs(ctx context.Context, q ethereum.Filter
 }
 
 func (client *client) SubscribeNewHead(ctx context.Context, ch chan<- *Head) (ethereum.Subscription, error) {
-	return client.pool.EthSubscribe(ctx, ch, "newHeads")
+	csf := newChainIDSubForwarder(client.ChainID(), ch)
+	err := csf.start(client.pool.EthSubscribe(ctx, csf.srcCh, "newHeads"))
+	if err != nil {
+		return nil, err
+	}
+	return csf, nil
+}
+
+var _ ethereum.Subscription = &chainIDSubForwarder{}
+
+// chainIDSubForwarder wraps a head subscription in order to intercept and augment each head with chainID before forwarding.
+type chainIDSubForwarder struct {
+	chainID *big.Int
+	destCh  chan<- *Head
+
+	srcCh  chan *Head
+	srcSub ethereum.Subscription
+
+	err   chan error
+	unSub chan chan struct{}
+}
+
+func newChainIDSubForwarder(chainID *big.Int, ch chan<- *Head) *chainIDSubForwarder {
+	return &chainIDSubForwarder{
+		chainID: chainID,
+		destCh:  ch,
+		srcCh:   make(chan *Head),
+		err:     make(chan error),
+		unSub:   make(chan chan struct{}),
+	}
+}
+
+// start spawns the forwarding loop for sub.
+func (c *chainIDSubForwarder) start(sub ethereum.Subscription, err error) error {
+	if err != nil {
+		close(c.srcCh)
+		return err
+	}
+	c.srcSub = sub
+	go c.forwardLoop()
+	return nil
+}
+
+// forwardLoop receives from src, adds the chainID, and then sends to dest.
+func (c *chainIDSubForwarder) forwardLoop() {
+	defer close(c.srcCh)
+	for {
+		select {
+		case err := <-c.srcSub.Err():
+			c.err <- err
+			return
+
+		case h := <-c.srcCh:
+			h.EVMChainID = utils.NewBig(c.chainID)
+			c.destCh <- h
```

Comment:
I'm wondering if this needs to select on `c.unSub` to avoid deadlock. Maybe the same for err above.

Reply:
We have a similar subscription forwarder in test code that needs work too, so I'm going to figure out that less critical case before revisiting this.

Reply:
Updated to yield to unsub during either operation. Will relocate to another file in next update.